### PR TITLE
Add multiple improvements and changes

### PR DESCRIPTION
**Pull request by 73sl4**

### CODE REVIEW

File: src/Payment.php [hunk 1]
```diff
@@ -36,13 +36,4 @@ public function capture($attributes = array())
 
         return $this->request('POST', $relativeUrl, $attributes);
     }
-
```
Comment (captn3m0) [inline src/Payment.php]: Why are we dropping this? Can't find how to do payment->refunds anymore

File: src/Request.php [hunk 2]
```diff
@@ -30,17 +30,23 @@ class Request
...
         );
 
-        $response = \Requests::request($url, self::$headers, $data, $method, $options);
+        $uaHeader = array(
+            'User-Agent' => $this->constructUa()
+        );
+
+        $headers = array_merge(self::$headers, $uaHeader);
+
+        $response = \Requests::request($url, $headers, $data, $method, $options);
```
Comment (captn3m0) [inline src/Request.php]: Put a use Requests instead

File: src/Request.php [hunk 3]
```diff
@@ -30,17 +30,23 @@ class Request
      */
     public function request($method, $url, $data = null)
     {
-        $url = Api::$baseUrl . $url;
+        $url = Api::getBaseUrl() . $url;
 
         if ($data === null)
             $data = array();
 
         $options = array(
```
Comment (captn3m0) [inline src/Request.php]: If we are doing a upgrade, let us drop support for EOL versions of PHP and start using square brackets?

File: src/Api.php [hunk 4]
```diff
@@ -4,46 +4,70 @@
...
-    public static $baseUrl = "https://api.razorpay.com/v1/";
+    protected static $baseUrl = 'https://api.razorpay.com/v1/';
 
-    public static $key = null;
+    protected static $key = null;
 
-    public static $secret = null;
+    protected static $secret = null;
+
+    public static $appInfo = null;
```
Comment (captn3m0) [inline src/Api.php]: Put a comment about how/why to use this

File: src/Utility.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public function verifyWebhookSignature($payload, $exceptedSignature)
+    {
+        return self::verifySignature($payload, $exceptedSignature);
+    }
+
+    public function verifySignature($payload, $exceptedSignature)
+    {
+        $actualSignature = hmac_hash(self::SHA256, $payload, Api::getSecret());
```
Comment (captn3m0) [inline src/Utility.php]: Function is called `hash_hmac`, not `hmac_hash`

File: src/Utility.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+        if (function_exists('hash_equals'))
+        {
+            if (hash_equals($actualSignature, $exceptedSignature) === true)
+            {
+                return true;
+            }
+        }
+        else if ($actualSignature === $exceptedSignature)
```
Comment (captn3m0) [inline src/Utility.php]: 1. If we enfoce 5.6, we can drop the function_exists check
2. If we keep the check, we need a constant time comparison here

File: src/Request.php (hunk 3 above)
Comment (73sl4) [inline src/Request.php]: Don't want to do it right now.

File: src/Utility.php [hunk 7]
```diff
@@ -33,12 +33,25 @@ public function verifySignature($payload, $exceptedSignature)
                 return true;
             }
         }
-        else if ($actualSignature === $exceptedSignature)
+        else if (strlen($exceptedSignature) === strlen($actualSignature))
         {
-            return true;
+            $res = $exceptedSignature ^ $actualSignature;
```
Comment (captn3m0) [inline src/Utility.php]: Maybe define as a private `hash_equals` instead?

File: src/Api.php (hunk 4 above)
Comment (73sl4) [inline src/Api.php]: Done.

File: src/Utility.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    public function verifyPaymentSignature($attributes)
+    {
+        $exceptedSignature = $attributes['razorpay_signature'];
+        $orderId = $attributes['razorpay_order_id'];
+        $paymentId = $attributes['razorpay_payment_id'];
+
+        $payload = $orderId . '|' . $paymentId;
+
+        return self::validateSignature($payload, $exceptedSignature);
```
Comment (captn3m0) [inline src/Utility.php]: @73sl4 validateSignature is not defined anywhere

File: src/Request.php (hunk 3 above)
Comment (shashankkumar) [inline src/Request.php]: We should not drop support for PHP 5.3 anytime soon.

File: src/Request.php [hunk 9]
```diff
@@ -164,4 +136,77 @@ protected function throwServerError($body, $httpStatusCode)
...
+        );
+
+        $headers = array_merge(self::$headers, $uaHeader);
+
+        return $headers;
+    }
+
+    protected function constructUa()
+    {
+        $ua = 'Razorpay/v1 PhpSdk/' . Api::VERSION;
```
Comment (captn3m0) [inline src/Request.php]: @shashankkumar I've been thinking of including PHP Version here as well, to help us gauge usage. Thoughts?

File: src/Api.php [hunk 10]
```diff
@@ -4,46 +4,93 @@
...
      */
-    function setHeader($header, $value)
+    public function setHeader($header, $value)
     {
         Request::addHeader($header, $value);
     }
 
+    public function setAppDetails($title, $version = null)
+    {
+        if ($appsDetails === null)
```
Comment (harman28) [inline src/Api.php]: Shouldn't this be `self:$appsDetails`?

File: src/Api.php [hunk 11]
```diff
@@ -4,46 +4,93 @@
...
     }
 
+    public function setAppDetails($title, $version = null)
+    {
+        if (self::$appsDetails === null)
+        {
+            self::$appsDetails = [];
+        }
+
+        $app = array(
```
Comment (gaurav1068) [inline src/Api.php]: use []

Comment (73sl4) [inline src/Api.php]: @gaurav1068 We are using array() to support older versions of php.

File: src/Errors/ErrorCode.php [hunk 12]
```diff
@@ -10,6 +10,8 @@ class ErrorCode
 
     public static function exists($code)
     {
+        $code = strtoupper($code);
+
         return defined(get_class().'::'.$code);
```
Comment (gaurav1068) [inline src/Errors/ErrorCode.php]: space after/before .

File: src/Invoice.php [hunk 13]
```diff
@@ -14,5 +15,10 @@ public function fetch($id)
     public function create($attributes = array())
     {
         return parent::create($attributes);
-    }    
+    }
+
+    public function all($options = array())
```
Comment (gaurav1068) [inline src/Invoice.php]: []

File: src/Request.php [hunk 14]
```diff
@@ -11,7 +12,6 @@
  */
 class Request
 {
-
```
Comment (gaurav1068) [inline src/Request.php]: does Razorpay-API need to change also ?

File: src/Request.php [hunk 15]
```diff
@@ -30,17 +30,21 @@ class Request
      */
     public function request($method, $url, $data = null)
```
Comment (gaurav1068) [inline src/Request.php]: take [] as default for data

File: src/Request.php [hunk 16]
```diff
@@ -85,31 +89,16 @@ protected function checkErrors($response)
             $this->throwServerError($body, $httpStatusCode);
         }
 
-        if ($httpStatusCode !== 200)
+        if (($httpStatusCode < 200) or
```
Comment (gaurav1068) [inline src/Request.php]: use some constants for 200 and 300

File: src/Request.php [hunk 17]
```diff
@@ -147,4 +136,77 @@ protected function throwServerError($body, $httpStatusCode)
...
+
+        foreach ($appsDetails as $app)
+        {
+            if ((isset($app['title'])) and (is_string($app['title'])))
+            {
+                $appUa = $app['title'];
+
+                if ((isset($app['version'])) and (is_scalar($app['version'])))
+                {
+                    $appUa .= '/'.$app['version'];
```
Comment (gaurav1068) [inline src/Request.php]: space after/before .

Comment (shashankkumar) [inline src/Request.php]: No need for space.

**Comment by gaurav1068** [comment]
tes cases are failing can you fix that

File: src/Invoice.php (hunk 13 above)
Comment (shashankkumar) [inline src/Invoice.php]: same as earlier. To support 5.3

File: src/Api.php [hunk 18]
```diff
@@ -4,46 +4,93 @@
...
+    public function setHeader($header, $value)
     {
         Request::addHeader($header, $value);
     }
 
+    public function setAppDetails($title, $version = null)
+    {
+        if (self::$appsDetails === null)
+        {
+            self::$appsDetails = [];
```
Comment (gaurav1068) [inline src/Api.php]: @73sl4 here we are using [], either use array or [] across codebase

File: src/Request.php (hunk 17 above)
Comment (captn3m0) [inline src/Request.php]: second `.` would need space

File: src/Api.php [hunk 19]
```diff
@@ -4,46 +4,93 @@
...
      *  Set Headers
      *
      */
-    function setHeader($header, $value)
+    public function setHeader($header, $value)
     {
         Request::addHeader($header, $value);
     }
 
+    public function setAppDetails($title, $version = null)
```
Comment (gaurav1068) [inline src/Api.php]: if setAppDetails is not called , then getAppDetailsUa will fail in Request.php as appsDetails is by default as null.
change default value to array()

File: src/Invoice.php [hunk 20]
```diff
@@ -1,10 +1,11 @@
 <?php
+
 namespace Razorpay\Api;
 
 class Invoice extends Entity
 {
-    /**
-     *  @param $id invoice
+    /*
```
Comment (gaurav1068) [inline src/Invoice.php]: add extra *

File: src/Request.php [hunk 21]
```diff
@@ -147,4 +136,77 @@ protected function throwServerError($body, $httpStatusCode)
...
+     * Verifies error is in proper format. If not then
+     * throws ServerErrorException
+     *
+     * @param  array $body
+     * @param  int $httpStatusCode
+     * @return void
+     */
+    protected function verifyErrorFormat($body, $httpStatusCode)
+    {
+        if (is_array($body) === false)
```
Comment (gaurav1068) [inline src/Request.php]: all the three of condition are doing the same thing, doesnt it make sense to have a single condition with or

Comment (73sl4) [inline src/Request.php]: It's separate because of readibility

File: src/Request.php (hunk 14 above)
Comment (73sl4) [inline src/Request.php]: Not really.

File: README.md [hunk 22]
```diff
@@ -38,8 +38,9 @@ $api = new Api($api_key, $api_secret);
 $api->payment->all($options); // Returns array of payment objects
 $payment = $api->payment->fetch($id); // Returns a particular payment
 $api->payment->fetch($id)->capture(array('amount'=>$amount)); // Captures a payment
-$api->payment->fetch($id)->refund(); // Refunds a payment 
-$api->payment->fetch($id)->refund(array('amount'=>$refundAmount)); // Partially refunds a payment
+$api->refund->create(array('payment_id' => $id)); // Creates refund for a payment
+$api->refund->create(array('payment_id' => $id, 'amount'=>$refundAmount)); // Creates partial refund for a payment
+$refund = $api->refund->fetch($refundId); // Returns a particular refund
```
Comment (harman28) [inline README.md]: This README needs to be updated with Orders API usage, Invoice usage, and also usage of the Utility class, which is not documented anywhere else.

**Pull request merged**